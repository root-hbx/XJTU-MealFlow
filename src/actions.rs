use color_eyre::eyre::Context;

use crate::{
    component::input::InputAction,
    libs::transactions::FilterOptions,
    page::{
        cookie_input::CookieInputAction, fetch::FetchingAction, help_popup::HelpPopupAction,
        transactions::TransactionAction,
    },
    utils::help_msg::HelpMsg,
};

#[derive(Clone, Debug)]
pub enum Action {
    Tick,
    Layer(LayerManageAction),
    SwitchInputMode(bool),

    Transaction(TransactionAction),
    Fetching(FetchingAction),
    CookieInput(CookieInputAction),
    HelpPopup(HelpPopupAction),

    Comp((CompAction, u64)),

    Quit,
    Render,

    #[cfg(test)]
    TestPage(crate::component::input::test::TestInputPageAction),
}
#[derive(Clone, Debug)]
pub enum Layers {
    #[allow(dead_code)]
    Home,
    Fetch,
    Transaction(Option<FilterOptions>),
    CookieInput,
    Help(HelpMsg),
}

#[derive(Clone, Debug)]
pub enum CompAction {
    Input(InputAction),
    #[allow(dead_code)]
    Placeholder,
}

#[derive(Clone, Debug)]
/// These actions should only be sent by page at the top of the stack
/// and should only be handled by the root app.
pub enum LayerManageAction {
    PushPage(Layers),
    SwapPage(Layers),
    PopPage,
}

impl From<LayerManageAction> for Action {
    fn from(value: LayerManageAction) -> Self {
        Action::Layer(value)
    }
}

#[derive(Clone, Debug)]
pub struct ActionSender(pub tokio::sync::mpsc::UnboundedSender<Action>);

impl ActionSender {
    pub fn send<T: Into<Action>>(&self, action: T) {
        self.0.send(action.into()).with_context(||"Action Receiver is dropped or closed, which should not happen if app is still running.").unwrap();
    }
}
impl From<tokio::sync::mpsc::UnboundedSender<Action>> for ActionSender {
    fn from(value: tokio::sync::mpsc::UnboundedSender<Action>) -> Self {
        ActionSender(value)
    }
}
