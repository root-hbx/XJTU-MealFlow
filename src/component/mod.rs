//! Component module defines the core trait for UI components in the application.
//! 
//! A Component is an *interactive* UI element that has its own actions but lives within a [`crate::page::Layer`].
//!
//! This module provides the base Component trait that all UI components must implement,
//! enabling consistent event handling and state updates across the application.

pub(crate) mod input;

use crate::{actions::Action, page::WidgetExt};
use color_eyre::eyre::Result;

/// A trait that defines the core functionality required for UI components.
///
/// Components are interactive UI elements that can:
/// - Handle input events
/// - Update their state based on actions
/// - Be uniquely identified
///
/// # Type Parameters
///
/// The trait requires implementing [`WidgetExt`] for basic widget functionality.
///
/// # Examples
///
/// ```rust
/// struct MyComponent {
///     id: u64,
///     // ... other fields
/// }
///
/// impl Component for MyComponent {
///     fn get_id(&self) -> u64 {
///         self.id
///     }
///     
///     fn handle_events(&self, event: &Event) -> Result<()> {
///         // Handle events...
///         Ok(())
///     }
///     
///     fn update(&mut self, action: &Action) -> Result<()> {
///         // Update component state...
///         Ok(())
///     }
/// }
/// ```
pub(crate) trait Component: WidgetExt {
    #[allow(dead_code)]
    fn get_id(&self) -> u64;

    fn handle_events(&self, event: &crate::tui::Event) -> Result<()>;

    fn update(&mut self, action: &Action) -> Result<()>;
}
