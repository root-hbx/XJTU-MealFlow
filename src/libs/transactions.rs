use std::{
    path::PathBuf,
    sync::{Arc, Mutex},
};

use chrono::{DateTime, Local};
use color_eyre::eyre::{Context, Result, bail};
use rusqlite::{Connection, params};

#[derive(Debug, Clone)]
pub struct Transaction {
    pub id: u64,
    pub time: DateTime<Local>,
    pub amount: f64,
    pub merchant: String,
}

#[derive(Debug, Clone)]
pub struct TransactionManager {
    conn: Arc<Mutex<Connection>>,
}

impl TransactionManager {
    pub fn new(db_path: Option<PathBuf>) -> Result<Self> {
        let conn = match db_path.as_ref() {
            Some(db_path) => {
                std::fs::create_dir_all(db_path.parent().unwrap())
                    .context("Failed to create dir for local cache DB")?;
                Connection::open(db_path.clone()).with_context(|| {
                    format!(
                        "Failed to open local cache DB at {}",
                        db_path.to_str().unwrap_or("INVALID PATH")
                    )
                })?
            }
            None => Connection::open_in_memory()?,
        };

        // Initialize the database
        TransactionManager::init_db(&conn)
            .with_context(|| "Failed to initialize local cache DB")?;

        Ok(TransactionManager {
            conn: Arc::new(Mutex::new(conn)),
        })
    }

    fn init_db(conn: &Connection) -> Result<(), rusqlite::Error> {
        conn.execute(
            "CREATE TABLE IF NOT EXISTS transactions (
                id INTEGER PRIMARY KEY,
                time TEXT NOT NULL,
                amount REAL NOT NULL,
                merchant TEXT NOT NULL
            )",
            [],
        )?;
        conn.execute(
            "CREATE TRIGGER IF NOT EXISTS prevent_transaction_conflict 
                BEFORE INSERT ON transactions
                FOR EACH ROW
                BEGIN
                    SELECT CASE
                    WHEN EXISTS (
                        SELECT 1 FROM transactions 
                        WHERE id = NEW.id 
                        AND time = NEW.time 
                        AND amount = NEW.amount 
                        AND merchant = NEW.merchant
                    ) THEN
                        RAISE(IGNORE)  -- 完全相同的记录则静默跳过
                    WHEN EXISTS (
                        SELECT 1 FROM transactions 
                        WHERE id = NEW.id
                    ) THEN
                        RAISE(ABORT, 'Conflict: Existing transaction with different data')  -- ID存在但数据不同时终止
                    END;
                END;",
            [],
        )?;
        conn.execute(
            "CREATE TABLE IF NOT EXISTS cookies (
            account TEXT PRIMARY KEY,
            cookie TEXT NOT NULL
        )",
            [],
        )?;
        Ok(())
    }

    pub fn insert(&self, transactions: &Vec<Transaction>) -> Result<()> {
        let conn = self.conn.lock().unwrap();

        // insert at once
        let mut stmt = conn
            .prepare("INSERT INTO transactions (id, time, amount, merchant) VALUES (?, ?, ?, ?)")?;

        for transaction in transactions {
            stmt.execute(params![
                transaction.id,
                transaction.time.to_rfc3339(),
                transaction.amount,
                transaction.merchant
            ])
            .with_context(|| {
                format!(
                    "Error when inserting transactions into Database, transaction: {:?}",
                    transaction
                )
            })?;
        }
        Ok(())
    }

    pub fn fetch_all(&self) -> Result<Vec<Transaction>, rusqlite::Error> {
        let conn = self.conn.lock().unwrap();
        let mut stmt = conn.prepare("SELECT id, time, amount, merchant FROM transactions")?;
        let transactions = stmt.query_map([], |row| {
            let time_str: String = row.get(1)?;
            let time = chrono::DateTime::parse_from_rfc3339(&time_str)
                .map_err(|e| rusqlite::Error::InvalidParameterName(e.to_string()))?
                .with_timezone(&Local);

            Ok(Transaction {
                id: row.get(0)?,
                time,
                amount: row.get(2)?,
                merchant: row.get(3)?,
            })
        })?;

        Ok(transactions.filter_map(|t| t.ok()).collect())
    }

    pub fn fetch_count(&self) -> Result<u64> {
        let conn = self.conn.lock().unwrap();
        let mut stmt = conn.prepare("SELECT COUNT(*) FROM transactions")?;
        let count: i64 = stmt.query_row([], |row| row.get(0))?;
        Ok(count as u64)
    }

    #[allow(dead_code)]
    pub fn clear_db(&self) -> Result<(), rusqlite::Error> {
        let conn = self.conn.lock().unwrap();
        conn.execute("DELETE FROM transactions", [])?;
        Ok(())
    }

    /// Update account(optional) and cookie in cookies table
    ///
    /// If there is already a record, update it. Otherwise, insert a new record.
    /// There should always be only one records in cookies table
    pub fn update_account(&self, account: &str) -> Result<()> {
        let conn = self.conn.lock().unwrap();

        // Check if there are existing records
        let existing = conn.query_row("SELECT cookie FROM cookies LIMIT 1", [], |row| {
            row.get::<_, String>(0)
        });

        // Determine account value to use
        let cookie = existing.unwrap_or_default();

        // Replace the record
        conn.execute("DELETE FROM cookies", [])?;
        let mut stmt = conn.prepare("INSERT INTO cookies (account, cookie) VALUES (?, ?)")?;
        stmt.execute(params![account, cookie])?;
        Ok(())
    }

    pub fn update_cookie(&self, cookie: &str) -> Result<()> {
        let conn = self.conn.lock().unwrap();

        // Check if there are existing records
        let existing = conn.query_row("SELECT account FROM cookies LIMIT 1", [], |row| {
            row.get::<_, String>(0)
        });

        // Determine account value to use
        let account = existing.unwrap_or_default();

        // Replace the record
        conn.execute("DELETE FROM cookies", [])?;
        let mut stmt = conn.prepare("INSERT INTO cookies (account, cookie) VALUES (?, ?)")?;
        stmt.execute(params![account, cookie])?;
        Ok(())
    }

    pub fn get_account_cookie(&self) -> Result<(String, String)> {
        let conn = self.conn.lock().unwrap();
        let mut stmt = conn.prepare("SELECT account, cookie FROM cookies")?;
        let mut rows = stmt.query([])?;
        let row = rows.next()?;
        match row {
            Some(row) => {
                let account: String = row.get(0)?;
                let cookie: String = row.get(1)?;
                if account.is_empty() || cookie.is_empty() {
                    bail!("Account or cookie is empty");
                }
                Ok((account, cookie))
            }
            None => bail!("No account and cookie found"),
        }
    }

    #[cfg(test)]
    pub fn get_account_cookie_may_empty(&self) -> Result<(String, String)> {
        let conn = self.conn.lock().unwrap();
        let mut stmt = conn.prepare("SELECT account, cookie FROM cookies")?;
        let mut rows = stmt.query([])?;
        let row = rows.next()?;
        match row {
            Some(row) => {
                let account: String = row.get(0)?;
                let cookie: String = row.get(1)?;
                Ok((account, cookie))
            }
            None => bail!("No account and cookie found"),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_transaction_manager() {
        let manager = TransactionManager::new(None).unwrap();

        manager.clear_db().unwrap();

        let transactions = vec![
            Transaction {
                id: 1,
                time: Local::now(),
                amount: -100.0,
                merchant: "Amazon".to_string(),
            },
            Transaction {
                id: 2,
                time: Local::now(),
                amount: -200.0,
                merchant: "Google".to_string(),
            },
        ];

        manager.insert(&transactions).unwrap();

        let fetched = manager.fetch_all().unwrap();
        assert_eq!(fetched.len(), 2);
        assert_eq!(fetched[0].id, 1);
        assert_eq!(fetched[0].amount, -100.0);
        assert_eq!(fetched[0].merchant, "Amazon");

        assert_eq!(fetched[1].id, 2);
        assert_eq!(fetched[1].amount, -200.0);
        assert_eq!(fetched[1].merchant, "Google");
    }

    #[test]
    fn test_account_cookie() {
        let manager = TransactionManager::new(None).unwrap();

        manager.update_account("test_account").unwrap();
        let (account, cookie) = manager.get_account_cookie_may_empty().unwrap();
        assert_eq!(account, "test_account");
        assert_eq!(cookie, "");

        manager.update_cookie("test_cookie").unwrap();
        let (account, cookie) = manager.get_account_cookie().unwrap();
        assert_eq!(account, "test_account");
        assert_eq!(cookie, "test_cookie");

        manager.update_account("test_account2").unwrap();
        let (account, cookie) = manager.get_account_cookie().unwrap();
        assert_eq!(account, "test_account2");
        assert_eq!(cookie, "test_cookie");
    }

    #[test]
    fn test_fetch_count() {
        let manager = TransactionManager::new(None).unwrap();
        manager.clear_db().unwrap();

        // Initially should have zero transactions
        let count = manager.fetch_count().unwrap();
        assert_eq!(count, 0);

        // Add some transactions
        let transactions = vec![
            Transaction {
                id: 1,
                time: Local::now(),
                amount: -100.0,
                merchant: "Amazon".to_string(),
            },
            Transaction {
                id: 2,
                time: Local::now(),
                amount: -200.0,
                merchant: "Google".to_string(),
            },
        ];

        manager.insert(&transactions).unwrap();

        // Should now have 2 transactions
        let count = manager.fetch_count().unwrap();
        assert_eq!(count, 2);

        // Add more transactions
        let more_transactions = vec![Transaction {
            id: 3,
            time: Local::now(),
            amount: -300.0,
            merchant: "Apple".to_string(),
        }];

        manager.insert(&more_transactions).unwrap();

        // Should now have 3 transactions
        let count = manager.fetch_count().unwrap();
        assert_eq!(count, 3);

        // Clear the database
        manager.clear_db().unwrap();

        // Should now have 0 transactions again
        let count = manager.fetch_count().unwrap();
        assert_eq!(count, 0);
    }

    #[test]
    fn multithread_access() {
        let manager = TransactionManager::new(None).unwrap();

        let transactions = vec![
            Transaction {
                id: 1,
                time: Local::now(),
                amount: -100.0,
                merchant: "Amazon".to_string(),
            },
            Transaction {
                id: 2,
                time: Local::now(),
                amount: -200.0,
                merchant: "Google".to_string(),
            },
        ];

        let manager_clone = manager.clone();
        std::thread::spawn(move || {
            manager_clone.insert(&transactions).unwrap();
        })
        .join()
        .unwrap();

        let fetched = manager.fetch_all().unwrap();
        assert_eq!(fetched.len(), 2);
        assert_eq!(fetched[0].id, 1);
    }
}
