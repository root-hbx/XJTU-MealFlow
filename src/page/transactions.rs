use std::cmp::max;

use crate::{
    actions::{Action, ActionSender, NaviTarget},
    libs::transactions::{Transaction, TransactionManager},
    tui::Event,
    utils::help_msg::{HelpEntry, HelpMsg},
};

use super::{EventLoopParticipant, Page, WidgetExt};
use color_eyre::eyre::Result;
use crossterm::event::KeyCode;
use lazy_static::lazy_static;
use ratatui::{
    Frame,
    layout::{Alignment, Constraint, Layout, Margin, Rect},
    style::{Color, Modifier, Style, Stylize, palette::tailwind},
    text::Text,
    widgets::{
        Cell, HighlightSpacing, Row, Scrollbar, ScrollbarOrientation, ScrollbarState, Table,
        TableState,
    },
};
use unicode_width::UnicodeWidthStr;

struct TableColors {
    buffer_bg: Color,
    header_bg: Color,
    header_fg: Color,
    row_fg: Color,
    selected_row_style_fg: Color,
    selected_column_style_fg: Color,
    selected_cell_style_fg: Color,
    normal_row_color: Color,
    alt_row_color: Color,
    // footer_border_color: Color,
}

impl Default for TableColors {
    fn default() -> Self {
        Self {
            buffer_bg: tailwind::SLATE.c950,
            header_bg: tailwind::INDIGO.c900,
            header_fg: tailwind::SLATE.c200,
            row_fg: tailwind::INDIGO.c200,
            selected_row_style_fg: tailwind::INDIGO.c400,
            selected_column_style_fg: tailwind::INDIGO.c400,
            selected_cell_style_fg: tailwind::INDIGO.c600,
            normal_row_color: tailwind::SLATE.c950,
            alt_row_color: tailwind::SLATE.c900,
            // footer_border_color: tailwind::INDIGO.c400,
        }
    }
}

lazy_static! {
    static ref TABLE_COLORS: TableColors = TableColors::default();
}

const ITEM_HEIGHT: usize = 3;

#[derive(Clone, Debug)]
pub struct Transactions {
    transactions: Vec<crate::libs::transactions::Transaction>,

    tx: crate::actions::ActionSender,
    manager: TransactionManager,

    table_state: TableState,
    scroll_state: ScrollbarState,
    longest_item_lens: (usize, usize, usize),
}

impl Transactions {
    pub fn new(tx: ActionSender, manager: TransactionManager) -> Self {
        Self {
            transactions: Default::default(),
            tx,
            manager,

            table_state: TableState::default(),
            scroll_state: ScrollbarState::default(),
            longest_item_lens: (0, 0, 0),
        }
    }

    fn get_help_msg(&self) -> HelpMsg {
        let help_msg: HelpMsg = vec![
            HelpEntry::new_plain("Move focus: hjkl"),
            HelpEntry::new('f', "Fetch"),
            HelpEntry::new('l', "Load from local cache"),
        ]
        .into();
        help_msg
    }
}

#[derive(Clone, Debug)]
pub enum TransactionAction {
    LoadTransactions,
    ChangeRowFocus(isize),
}

impl From<TransactionAction> for Action {
    fn from(val: TransactionAction) -> Self {
        Action::Transaction(val)
    }
}

impl WidgetExt for Transactions {
    fn render(&mut self, frame: &mut Frame, area: Rect) {
        let vertical = &Layout::vertical([Constraint::Fill(1), Constraint::Length(3)]);
        let rects = vertical.split(area);

        self.render_table(frame, rects[0]);
        self.render_scrollbar(frame, rects[0]);

        self.get_help_msg().render(frame, rects[1]);
    }
}

impl EventLoopParticipant for Transactions {
    fn handle_events(&self, event: Event) -> Result<()> {
        if let Event::Key(key) = event {
            match (key.modifiers, key.code) {
                // navigate to fetch page
                (_, KeyCode::Char('f')) => self.tx.send(Action::NavigateTo(NaviTarget::Fetch)),
                (_, KeyCode::Char('l')) => self.tx.send(TransactionAction::LoadTransactions),
                (_, KeyCode::Char('j')) => self.tx.send(TransactionAction::ChangeRowFocus(1)),
                (_, KeyCode::Char('k')) => self.tx.send(TransactionAction::ChangeRowFocus(-1)),
                _ => (),
            }
        };
        Ok(())
    }

    fn update(&mut self, action: Action) {
        if let Action::Transaction(action) = action {
            match action {
                TransactionAction::LoadTransactions => {
                    self.transactions = self.manager.fetch_all().unwrap();
                    self.transactions.sort_by(|a, b| b.time.cmp(&a.time));
                    self.scroll_state = self
                        .scroll_state
                        .content_length(self.transactions.len() * ITEM_HEIGHT);
                    self.longest_item_lens = constraint_len_calculator(&self.transactions);
                }
                TransactionAction::ChangeRowFocus(index) => {
                    let cur_index = self.table_state.selected().unwrap_or(0);
                    let max = self.transactions.len();
                    if max == 0 {
                        return;
                    }
                    // add index to current index, wrap around from 0  to max
                    let new_index = if index < 0 {
                        (cur_index as isize + index + max as isize) as usize % max
                    } else {
                        (cur_index as isize + index) as usize % max
                    };
                    self.table_state.select(Some(new_index));
                    self.scroll_state = self.scroll_state.position(new_index * ITEM_HEIGHT);
                }
            }
        }
    }
}

impl Page for Transactions {
    fn get_name(&self) -> String {
        "Transactions".to_string()
    }

    fn init(&mut self) {
        self.tx.send(TransactionAction::LoadTransactions)
    }
}

impl Transactions {
    fn render_table(&mut self, frame: &mut Frame, area: Rect) {
        let header_style = Style::default()
            .fg(TABLE_COLORS.header_fg)
            .bg(TABLE_COLORS.header_bg);
        let selected_row_style = Style::default()
            .add_modifier(Modifier::REVERSED)
            .fg(TABLE_COLORS.selected_row_style_fg);
        let selected_col_style = Style::default().fg(TABLE_COLORS.selected_column_style_fg);
        let selected_cell_style = Style::default()
            .add_modifier(Modifier::REVERSED)
            .fg(TABLE_COLORS.selected_cell_style_fg);

        let header = ["金额", "时间", "商家"]
            .into_iter()
            .map(Cell::from)
            .collect::<Row>()
            .style(header_style)
            .height(1);

        let rows = self.transactions.iter().enumerate().map(|(i, t)| {
            let color = match i % 2 {
                0 => TABLE_COLORS.normal_row_color,
                _ => TABLE_COLORS.alt_row_color,
            };
            Row::new(
                vec![
                    Text::from(format!("\n{}\n", t.amount)).alignment(Alignment::Right),
                    Text::from(format!("\n{}\n", t.time)),
                    Text::from(format!("\n{}\n", t.merchant)),
                ]
                .into_iter(),
            )
            .style(Style::new().fg(TABLE_COLORS.row_fg).bg(color))
            .height(ITEM_HEIGHT as u16)
        });
        let bar = " █ ";

        let t = Table::new(
            rows,
            [
                Constraint::Length((self.longest_item_lens.0 + 2).try_into().unwrap()),
                Constraint::Min((self.longest_item_lens.1 + 2).try_into().unwrap()),
                Constraint::Min((self.longest_item_lens.2 + 2).try_into().unwrap()),
            ],
        )
        .header(header)
        .row_highlight_style(selected_row_style)
        .column_highlight_style(selected_col_style)
        .cell_highlight_style(selected_cell_style)
        .highlight_symbol(Text::from(vec!["".into(), bar.into(), "".into()]))
        .bg(TABLE_COLORS.buffer_bg)
        .highlight_spacing(HighlightSpacing::Always)
        .column_spacing(4);

        frame.render_stateful_widget(t, area, &mut self.table_state);
    }

    fn render_scrollbar(&mut self, frame: &mut Frame, area: Rect) {
        frame.render_stateful_widget(
            Scrollbar::default()
                .orientation(ScrollbarOrientation::VerticalRight)
                .begin_symbol(None)
                .end_symbol(None),
            area.inner(Margin {
                vertical: 1,
                horizontal: 1,
            }),
            &mut self.scroll_state,
        );
    }
}

fn constraint_len_calculator(items: &Vec<Transaction>) -> (usize, usize, usize) {
    items.iter().fold((0, 0, 0), |acc, item| {
        let amount_len = max(
            acc.0,
            UnicodeWidthStr::width(item.amount.to_string().as_str()),
        );
        let time_len = max(
            acc.1,
            UnicodeWidthStr::width(item.time.to_string().as_str()),
        );
        let merchant_len = max(
            acc.2,
            UnicodeWidthStr::width_cjk(item.merchant.to_string().as_str()),
        );
        (amount_len, time_len, merchant_len)
    })
}

#[cfg(test)]
mod test {
    use crate::libs::fetcher;

    use super::*;
    use insta::assert_snapshot;
    use ratatui::{Terminal, backend::TestBackend};
    use tokio::sync::mpsc::{self, UnboundedReceiver};

    fn get_test_objs() -> (UnboundedReceiver<Action>, Transactions) {
        let (tx, mut rx) = mpsc::unbounded_channel::<Action>();

        let manager = TransactionManager::new(None).unwrap();
        let data = fetcher::test_utils::get_mock_data(50);
        manager.insert(&data).unwrap();

        let mut transaction = Transactions::new(tx.into(), manager);
        transaction.init();

        while let Ok(action) = rx.try_recv() {
            transaction.update(action);
        }

        assert!(!transaction.transactions.is_empty());

        (rx, transaction)
    }

    #[test]
    fn order() {
        let (_, transaction) = get_test_objs();
        transaction
            .transactions
            .iter()
            .enumerate()
            .for_each(|(i, t)| {
                if i == 0 {
                    return;
                }
                assert!(t.time <= transaction.transactions[i - 1].time);
            });
    }

    #[test]
    fn navigation() {
        let (mut rx, mut transaction) = get_test_objs();
        assert_eq!(transaction.table_state.selected(), None);

        transaction.event_loop_once(&mut rx, 'j'.into());
        assert_eq!(transaction.table_state.selected(), Some(1));

        transaction.event_loop_once(&mut rx, 'k'.into());
        assert_eq!(transaction.table_state.selected(), Some(0));

        transaction.event_loop_once(&mut rx, 'k'.into());
        assert_eq!(
            transaction.table_state.selected(),
            Some(transaction.transactions.len() - 1)
        );
    }

    #[test]
    fn render() {
        let (mut rx, mut transaction) = get_test_objs();
        let mut terminal = Terminal::new(TestBackend::new(80, 25)).unwrap();

        terminal
            .draw(|frame| transaction.render(frame, frame.area()))
            .unwrap();
        assert_snapshot!(terminal.backend());

        transaction.event_loop_once(&mut rx, 'k'.into());
        terminal
            .draw(|frame| transaction.render(frame, frame.area()))
            .unwrap();
        assert_snapshot!(terminal.backend());
    }
}
