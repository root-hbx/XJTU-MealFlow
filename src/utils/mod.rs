pub(crate) mod help_msg;
pub(crate) mod key_events;
