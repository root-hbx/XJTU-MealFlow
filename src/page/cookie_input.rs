use crossterm::event::KeyCode;
use ratatui::layout::{Constraint, Layout};

use crate::actions::Action;
use crate::actions::ActionSender;
use crate::actions::NaviTarget;
use crate::component::Component;
use crate::component::input::{InputComp, InputMode};
use crate::libs::transactions::TransactionManager;
use crate::utils::help_msg::{HelpEntry, HelpMsg};

use super::Page;

#[derive(Clone, Debug)]
pub struct CookieInput {
    state: Focus,
    manager: TransactionManager,

    input_mode: bool,
    tx: ActionSender,

    cookie_input: InputComp,
    account_input: InputComp,
}

impl CookieInput {
    pub fn new(action_tx: ActionSender, manager: TransactionManager, input_mode: bool) -> Self {
        let (account, cookie) = manager.get_account_cookie().unwrap_or_default();
        Self {
            state: Default::default(),
            manager,
            input_mode,
            cookie_input: InputComp::new(1, input_mode, action_tx.clone())
                .init_text(cookie)
                .title("Cookie"),
            account_input: InputComp::new(2, input_mode, action_tx.clone())
                .init_text(account)
                .title("Account"),
            tx: action_tx,
        }
    }

    pub fn get_help_msg(&self) -> crate::utils::help_msg::HelpMsg {
        let help_msg: HelpMsg = vec![
            HelpEntry::new_plain("Move focus: hjkl"),
            HelpEntry::new(KeyCode::Esc, "Back"),
        ]
        .into();
        match self.state {
            Focus::Account => help_msg.extend_ret(&self.account_input.get_help_msg()),
            Focus::Cookie => help_msg.extend_ret(&self.cookie_input.get_help_msg()),
        }
    }
}

#[derive(Default, Clone, Debug)]
pub enum Focus {
    #[default]
    Account,
    Cookie,
}

impl Focus {
    fn next(&self) -> Focus {
        match self {
            Focus::Account => Focus::Cookie,
            Focus::Cookie => Focus::Account,
        }
    }
    fn prev(&self) -> Focus {
        match self {
            Focus::Account => Focus::Cookie,
            Focus::Cookie => Focus::Account,
        }
    }
}

#[derive(Clone, Debug)]
pub(crate) enum CookieInputAction {
    ChangeState(Focus),
}

impl From<CookieInputAction> for Action {
    fn from(val: CookieInputAction) -> Self {
        Action::CookieInput(val)
    }
}

impl Page for CookieInput {
    fn render(&self, frame: &mut ratatui::Frame) {
        let chunks = &Layout::default()
            .constraints([Constraint::Fill(1), Constraint::Length(3)])
            .split(frame.area());

        let sub_chunks = &Layout::default()
            .margin(1)
            .constraints([Constraint::Length(5), Constraint::Length(5)])
            .split(chunks[0]);

        self.account_input.draw(frame, &sub_chunks[0]);
        self.cookie_input.draw(frame, &sub_chunks[1]);

        self.get_help_msg().render(frame, chunks[1]);
    }

    fn handle_events(&self, event: crate::tui::Event) -> color_eyre::eyre::Result<()> {
        if let crate::tui::Event::Key(key) = &event {
            if !self.input_mode {
                match (key.modifiers, key.code) {
                    (_, KeyCode::Char('k')) => self
                        .tx
                        .send(CookieInputAction::ChangeState(self.state.prev())),
                    (_, KeyCode::Char('j')) => self
                        .tx
                        .send(CookieInputAction::ChangeState(self.state.next())),
                    (_, KeyCode::Esc) => self.tx.send(Action::NavigateTo(NaviTarget::Fetch)),
                    _ => (),
                }
            }
        };
        self.account_input.handle_events(&event)?;
        self.cookie_input.handle_events(&event)?;
        Ok(())
    }

    fn update(&mut self, action: crate::actions::Action) {
        if let Action::SwitchInputMode(mode) = &action {
            self.input_mode = *mode;
        }

        if let Action::CookieInput(CookieInputAction::ChangeState(next_state)) = &action {
            self.state = next_state.clone();

            self.tx.send(self.account_input.get_switch_mode_action(
                if matches!(self.state, Focus::Account) {
                    InputMode::Focused
                } else {
                    InputMode::Idle
                },
            ));

            self.tx.send(self.cookie_input.get_switch_mode_action(
                if matches!(self.state, Focus::Cookie) {
                    InputMode::Focused
                } else {
                    InputMode::Idle
                },
            ));
        }

        if let Some(string) = self.account_input.parse_submit_action(&action) {
            self.manager.update_account(&string).unwrap();
        };
        if let Some(string) = self.cookie_input.parse_submit_action(&action) {
            self.manager.update_cookie(&string).unwrap();
        }

        self.account_input.update(&action).unwrap();
        self.cookie_input.update(&action).unwrap();
    }

    fn get_name(&self) -> String {
        "Cookie Input".to_string()
    }
    fn init(&mut self) {
        self.tx.send(
            self.account_input
                .get_switch_mode_action(InputMode::Focused),
        );
    }
}

#[cfg(test)]
mod test {

    use insta::assert_snapshot;
    use ratatui::backend::TestBackend;
    use tokio::sync::mpsc::{self, UnboundedReceiver};

    use super::*;
    use crate::tui::Event;
    use crate::utils::key_events::test_utils::{get_char_evt, get_key_evt};

    fn get_test_objs() -> (UnboundedReceiver<Action>, CookieInput) {
        let (tx, mut rx) = mpsc::unbounded_channel();
        let mut page = CookieInput::new(
            tx.clone().into(),
            TransactionManager::new(None).unwrap(),
            false,
        );
        page.init();
        while let Ok(action) = rx.try_recv() {
            page.update(action);
        }
        (rx, page)
    }

    #[test]
    fn test_navigation() {
        let (mut rx, mut page) = get_test_objs();
        assert!(matches!(page.state, Focus::Account));
        assert!(matches!(page.account_input.get_mode(), InputMode::Focused));

        page.event_loop_once(&mut rx, get_key_evt(KeyCode::Char('j')));
        assert!(matches!(page.state, Focus::Cookie));

        page.event_loop_once(&mut rx, get_key_evt(KeyCode::Char('j')));
        assert!(matches!(page.state, Focus::Account));

        page.event_loop_once(&mut rx, get_key_evt(KeyCode::Char('k')));
        assert!(matches!(page.state, Focus::Cookie));

        page.event_loop_once(&mut rx, get_key_evt(KeyCode::Char('k')));
        assert!(matches!(page.state, Focus::Account));
    }

    #[test]
    fn test_input_mode_change() {
        let (mut rx, mut page) = get_test_objs();
        assert!(!page.input_mode);
        page.event_loop_once_with_action(&mut rx, Action::SwitchInputMode(true));
        assert!(page.input_mode);
        page.event_loop_once_with_action(&mut rx, Action::SwitchInputMode(false));
        assert!(!page.input_mode);
    }

    #[test]
    fn test_account_input() {
        let (mut rx, mut page) = get_test_objs();

        page.event_loop_once(&mut rx, get_key_evt(KeyCode::Enter));
        assert!(page.input_mode);
        page.event_loop_once(&mut rx, get_char_evt('a'));
        page.event_loop_once(&mut rx, get_char_evt('j'));
        page.event_loop_once(&mut rx, get_key_evt(KeyCode::Enter));
        assert_eq!(page.manager.get_account_cookie_may_empty().unwrap().0, "aj");

        page.event_loop_once(&mut rx, get_key_evt(KeyCode::Enter));
        page.event_loop_once(&mut rx, get_key_evt(KeyCode::Left));
        page.event_loop_once(&mut rx, Event::Paste("kl".into()));
        page.event_loop_once(&mut rx, get_key_evt(KeyCode::Enter));
        assert_eq!(
            page.manager.get_account_cookie_may_empty().unwrap().0,
            "aklj"
        );
    }

    #[test]
    fn test_cookie_input() {
        let (mut rx, mut page) = get_test_objs();

        page.event_loop_once(&mut rx, get_char_evt('j'));
        page.event_loop_once(&mut rx, get_key_evt(KeyCode::Enter));
        assert!(page.input_mode);
        page.event_loop_once(&mut rx, get_char_evt('a'));
        page.event_loop_once(&mut rx, get_char_evt('j'));
        page.event_loop_once(&mut rx, get_key_evt(KeyCode::Enter));
        page.event_loop_once(&mut rx, get_key_evt(KeyCode::Esc));
        assert_eq!(page.manager.get_account_cookie_may_empty().unwrap().1, "aj");
    }

    #[test]
    fn test_cookie_input_render() {
        let (_, page) = get_test_objs();
        let mut terminal = ratatui::Terminal::new(TestBackend::new(80, 20)).unwrap();

        terminal
            .draw(|f| {
                page.render(f);
            })
            .unwrap();

        assert_snapshot!(terminal.backend())
    }
}
