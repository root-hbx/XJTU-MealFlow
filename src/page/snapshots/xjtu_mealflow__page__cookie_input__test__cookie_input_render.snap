---
source: src/page/cookie_input.rs
expression: terminal.backend()
---
"                                                                                "
" ╭Account─────────────────────────────────────────────────────────────────────╮ "
" │                                                                            │ "
" │                                                                            │ "
" │                                                                            │ "
" ╰────────────────────────────────────────────────────────────────────────────╯ "
" ╭Cookie──────────────────────────────────────────────────────────────────────╮ "
" │                                                                            │ "
" │                                                                            │ "
" │                                                                            │ "
" ╰────────────────────────────────────────────────────────────────────────────╯ "
"                                                                                "
"                                                                                "
"                                                                                "
"                                                                                "
"                                                                                "
"                                                                                "
"╭──────────────────────────────────────────────────────────────────────────────╮"
"│ Move focus: hjkl | Help: ? | Back: esc | Start input: enter                  │"
"╰──────────────────────────────────────────────────────────────────────────────╯"
